### PR TITLE
EnableWhenExpression context literal fhirpath supplement implementation

### LINKED ISSUE #1958: Support fhirpath supplement context literal
**Is your feature request related to a problem? Please describe.**
To make question fhirpath dynamic and have current questionnaire item we need to support fhirpath supplement context literal. This allow to get current questionnaire item as %context into fhirpath  expressions on questionnaire items. Details for spec defined here https://build.fhir.org/ig/HL7/sdc/expressions.html#fhirpath-supplements . 

**Describe the solution you'd like**
An implementation exists in PR #1957 . This has a catalog implementation as an example as well. The enableWhen evaluate method passes the questionnaireResponseItem when calling evaluate for expression. 

**Additional context**
This is a bug fix and the changes MAY affect existing use of enableWhenExpression where anyone tried to write expression for questionnaireResponse as $this. $this refers to current item under consideration aka %context as well. Old implementation might have entertained $this as questionnaireResponse. Hence anyone using expression wrongfully can have erratic behavior for their current enableWhenExpressions 

**Would you like to work on the issue?**
Done in PR #1957

### PR DESCRIPTION
**Pull request by maimoonak**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1958 

**Description**
To make question fhirpath dynamic and have current questionnaire item we need to support fhirpath supplement context literal. This allow to get current questionnaire item as %context into fhirpath expressions on questionnaire items. Details for spec defined here https://build.fhir.org/ig/HL7/sdc/expressions.html#fhirpath-supplements .
This is a bug fix and the changes MAY affect existing use of enableWhenExpression where anyone tried to write expression for questionnaireResponse as $this. $this refers to current item under consideration aka %context as well. Old implementation might have entertained $this as questionnaireResponse. Hence anyone using expression wrongfully can have erratic behavior for their current enableWhenExpressions

**Alternative(s) considered**
Bug fix

**Type**
Choose one: Bug fix

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/FhirPathUtil.kt [hunk 1]
```diff
@@ -34,3 +36,21 @@ internal val fhirPathEngine: FHIRPathEngine =
  */
 internal fun evaluateToDisplay(expressions: List<String>, data: Resource) =
   expressions.joinToString(" ") { fhirPathEngine.evaluateToString(data, it) }
+
+/**
+ * Evaluates the expressions over list of resources [QuestionnaireResponse] and
+ * [QuestionnaireResponseItemComponent] and returns the resulting elements FhirPath supplements
+ * https://build.fhir.org/ig/HL7/sdc/expressions.html#fhirpath-supplements %resource =
+ * [QuestionnaireResponse] %context = [QuestionnaireResponseItemComponent]
+ */
```
Comment (santosh-pingle) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/FhirPathUtil.kt:45]: nit: Could you please review the Kotlin API documentation one more time? I'm sorry, but I'm having trouble understanding it.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
Thanks @maimoonak for this PR.

I'm having some trouble understanding what this PR is doing beyond wrapping fhir path engine functions in wrappers.

To support fhirpath supplements, we need to use `FHIRPathEngineHostServices` so that it can resolve `%resource`, `%context` etc. It is the service `FHIRPathEngine` uses when it sees a variable with `%` that it cannot resolve.

**Comment by jingtang10** [comment]
@omarismail94 fyi - we discussed this before.

**Comment by maimoonak** [comment]
> Thanks @maimoonak for this PR.
> 
> I'm having some trouble understanding what this PR is doing beyond wrapping fhir path engine functions in wrappers.
> 
There is a bug where we are not passing focusResource (which is recognized as %resource in expression) and baseResource (which is recognized as %context or $this in expression) separately into evaluate function when evaluating enableWhen-expression. This leads to a missing feature i.e. use of %context to get current questionnaireResponseItem in expression as defined here https://build.fhir.org/ig/HL7/sdc/expressions.html#fhirpath-supplements and a bug where user can use %resource, %context, $this all constants for questionnaireResponse which is wrong.

> To support fhirpath supplements, we need to use `FHIRPathEngineHostServices` so that it can resolve `%resource`, `%context` etc. It is the service `FHIRPathEngine` uses when it sees a variable with `%` that it cannot resolve.

As per implementation here org.hl7.fhir.r4.utils.FHIRPathEngine#resolveConstant(ExecutionContext, String, boolean, ExpressionNode), the constants which are part of default fhir spec (https://hl7.org/fhir/fhirpath.html#variables and https://hl7.org/fhir/fhirpath.html#vars) i.e. %constant_name are resolved automatically by engine. We can not provide implementation for those. If any variable is not reserved one and is provided by specific IG, then we need to provide it via HostServices. Some of the examples of such cases are StructureMap extraction and SDC variable resolution. 

See attached image to see the code flow.

Screenshot from 2023-04-26 16-59-36

File: catalog/src/main/res/values/strings.xml [hunk 2]
```diff
@@ -43,6 +43,9 @@
     <string
         name="behavior_name_skip_logic_info"
     >If Yes is selected, a follow-up question is displayed. If No is selected, no follow-up questions are displayed.</string>
+    <string
+        name="behavior_name_enable_when_expression"
+    >Enable When Expression</string>
```
Comment (jingtang10) [inline catalog/src/main/res/values/strings.xml:48]:
```suggestion
    >Context Variables</string>
```

can you also rename the other elements...

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/FhirPathUtil.kt [hunk 3]
```diff
@@ -34,3 +36,22 @@ internal val fhirPathEngine: FHIRPathEngine =
  */
 internal fun evaluateToDisplay(expressions: List<String>, data: Resource) =
   expressions.joinToString(" ") { fhirPathEngine.evaluateToString(data, it) }
+
```
Comment (jingtang10) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/FhirPathUtil.kt:39]: we'll need to make changes to `evaluateToDisplay` as well i think

Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/FhirPathUtil.kt:39]: also change `ExpressionEvaluator`

Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/FhirPathUtil.kt:39]: we need to review the usage of fhir path engine in the whole of sdc, and make sure we're passing the right context in the rigth places, and hopefully creating a single place where we're passing these contexts.

but - i understand you might need this feature soon - and that can be done as a follow-up

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/FhirPathUtil.kt (hunk 3 above)
Comment (maimoonak) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/fhirpath/FhirPathUtil.kt:39]: New issue created here https://github.com/google/android-fhir/issues/1993 to record changes needed.

**Review by jingtang10 — APPROVED** [review]

**Pull request merged**